### ISSUE TITLE
libDeviceCollectorLibrary.a has invalid debug info

### ISSUE THREAD

**Issue by keith** [braintree/braintree_ios#539]
On 50d28cca423cd5f2e96b3d5937b83c44f100c305 `libDeviceCollectorLibrary.a` has invalid debug info with Xcode 12:

```
DIBlockByRefStruct on DICompositeType is no longer supported
!1229 = !DICompositeType(tag: DW_TAG_structure_type, scope: !3, file: !3, size: 384, flags: DIFlagReservedBit4, elements: !1230)
DIBlockByRefStruct on DICompositeType is no longer supported
!1246 = !DICompositeType(tag: DW_TAG_structure_type, scope: !3, file: !3, size: 256, flags: DIFlagReservedBit4, elements: !1247)
DIBlockByRefStruct on DICompositeType is no longer supported
!1251 = !DICompositeType(tag: DW_TAG_structure_type, scope: !3, file: !3, size: 384, flags: DIFlagReservedBit4, elements: !1252)
DIBlockByRefStruct on DICompositeType is no longer supported
!1256 = !DICompositeType(tag: DW_TAG_structure_type, scope: !3, file: !3, size: 384, flags: DIFlagReservedBit4, elements: !1257)
DIBlockByRefStruct on DICompositeType is no longer supported
!1261 = !DICompositeType(tag: DW_TAG_structure_type, scope: !3, file: !3, size: 384, flags: DIFlagReservedBit4, elements: !1262)
ld: warning: ignoring invalid debug info in BraintreeDataCollector/Kount/libDeviceCollectorLibrary.a(KDataCollector.o)
ld: fatal warning(s) induced error (-fatal_warnings)
```

It appears to be caused by this change https://reviews.llvm.org/D67453 and I assume it's because this library was built with an old version of Xcode. This surfaces when you try to produce bitcode while linking this library. Here's an example command:

```
/Applications/Xcode-12.0.0-beta6.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld \
 -dynamic \
 -dylib \
 -bitcode_verify \
 -bitcode_hide_symbols \
 -bitcode_symbol_map /tmp/foo BraintreeDataCollector/Kount/libDeviceCollectorLibrary.a \
 -platform_version ios 11.0.0 14.0 \
 -bitcode_bundle \
 -arch arm64 \
 -fatal_warnings \
 -all_load \
 -lSystem \
 -lobjc \
 -syslibroot /Applications/Xcode-12.0.0-beta6.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS14.0.sdk
```

(Normally this would happen when linking an app, but this reduces it for debugging)

**Comment by keith** [comment]
Looks like it's probably because this library hasn't been rebuilt / updated here for 3+ years

**Comment by tfulton** [comment]
Hi BT.  Created zendesk issue in following https://braintree.slack.com/archives/C02CXD2LC/p1600133635205500 

The above is going to be a blocker for merchant team as they plan out upgrade to Xcode 12.  Some relevant info:
* BT SDK version is 4.33.0
* Merchant attempting Beta 6 of Xcode 12 (didn't try earlier versions)

**Comment by scannillo** [comment]
Hello - we are looking into this issue

**Comment by scannillo** [comment]
Hi @keith - can I ask what command you're running that confirms this isn't an issue in Xcode 11?

**Comment by keith** [comment]
The same command as above with Xcode 11.5 doesn't produce this warning:

```
/Applications/Xcode-11.5.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld \
 -dynamic \
 -dylib \
 -bitcode_verify \
 -bitcode_hide_symbols \
 -bitcode_symbol_map /tmp/foo BraintreeDataCollector/Kount/libDeviceCollectorLibrary.a \
 -platform_version ios 11.0.0 14.0 \
 -bitcode_bundle \
 -arch arm64 \
 -fatal_warnings \
 -all_load \
 -lSystem \
 -lobjc \
 -syslibroot /Applications/Xcode-11.5.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk
```

Same with Xcode 11.7. Also our actual repro case of building our app with bitcode doesn't have this warning on 11.x either, we just discovered it testing Xcode 12

**Comment by scannillo** [comment]
@keith - we've filed an issue with our device data collector service, Kount. Feel free to follow along here: https://github.com/Kount/kount-ios-sdk/issues/21

**Comment by keith** [comment]
It looks like they have newer versions than the one from 2017 you're using, can you update to see if it fixes the issue? https://github.com/Kount/kount-ios-sdk/commits/master

**Comment by scannillo** [comment]
I did try those newer versions - they did not fix the issue

**Comment by yeradis** [comment]
This is happening on Xcode 12 GM too

**Comment by scannillo** [comment]
Hi @keith - I just created a sample app that includes the Braintree framework (and Kount v3) via the Xcode 12 GM. I uploaded it to TestFlight and am able to run the app on my phone just fine. So this warning doesn't seem to be a blocker.

Please let us know if this warning actually ends up blocking your ability to deploy your app.

**Comment by keith** [comment]
Correct it does not block us from shipping our app. But we don't want to ignore warnings here since it opens the door for new linker issues sneaking in until these are fixed.

**Comment by sestevens** [comment]
Hi @keith, we just released a [beta version](https://github.com/braintree/braintree_ios/releases/tag/5.0.0-beta1) of the SDK that includes a new version of this static library (Kount) where this issue is resolved. I'll leave this issue open until these changes are in GA.

Resources:
- [SPM integration instructions](https://github.com/braintree/braintree_ios/tree/5.0.0-beta1#swift-package-manager-v5-beta)
- [Version 5 migration guide](https://github.com/braintree/braintree_ios/blob/5.0.0-beta1/V5_MIGRATION.md)

**Comment by keith** [comment]
Thanks. Unfortunately we still support iOS 11 so we cannot upgrade

**Comment by sestevens** [comment]
This has been fixed in [version 5.0.0](https://github.com/braintree/braintree_ios/releases/tag/5.0.0). Unfortunately, we can't backport this fix to version 4.x.

When you're able to upgrade, please see our [migration guide](https://github.com/braintree/braintree_ios/blob/5.0.0/V5_MIGRATION.md) for steps.
